indexer/src/store.rs

```diff
@@ -0,0 +1,1298 @@
+        if lock_hash_index_states.is_empty() {
+            return;
+        }
+
+        // retains the lock hashes on fork chain and detach blocks
+        lock_hash_index_states.retain(|_, index_state| {
+            self.shared
+                .store()
+                .get_block_number(&index_state.block_hash)
+                != Some(index_state.block_number)
+        });
+        lock_hash_index_states
+            .iter()
+            .for_each(|(lock_hash, index_state)| {
+                let mut index_lock_hashes = HashSet::new();
+                index_lock_hashes.insert(lock_hash.to_owned());
+
+                let mut block = self
+                    .shared
+                    .store()
+                    .get_block(&index_state.block_hash)
+                    .expect("block exists");
+                // detach blocks until reach a block on main chain
+                self.commit_batch(|batch| {
+                    self.detach_block(batch, &index_lock_hashes, &block);
+                    while self
+                        .shared
+                        .store()
+                        .get_block_hash(block.header().number() - 1)
+                        != Some(block.header().parent_hash().to_owned())
+                    {
+                        block = self
+                            .shared
+                            .store()
+                            .get_block(block.header().parent_hash())
+                            .expect("block exists");
+                        self.detach_block(batch, &index_lock_hashes, &block);
+                    }
+                    let index_state = LockHashIndexState {
+                        block_number: block.header().number() - 1,
+                        block_hash: block.header().parent_hash().to_owned(),
+                    };
+                    batch.insert_lock_hash_index_state(lock_hash, &index_state);
+                });
+            });
+
+        // attach blocks until reach tip or batch limit
+        let mut lock_hash_index_states = self.get_lock_hash_index_states();
+        let min_block_number: BlockNumber = lock_hash_index_states
+            .values()
+            .min_by_key(|index_state| index_state.block_number)
+            .expect("none empty index states")
+            .block_number;
+
+        let (tip_number, tip_hash) = {
+            let tip_header = self
+                .shared
+                .store()
+                .get_tip_header()
+                .expect("tip header exists");
+            (tip_header.number(), tip_header.hash().to_owned())
+        };
+        self.commit_batch(|batch| {
+            (min_block_number + 1..=tip_number)
+                .take(BATCH_ATTACH_BLOCK_NUMS)
+                .for_each(|block_number| {
+                    let index_lock_hashes = lock_hash_index_states
+                        .iter()
+                        .filter(|(_, index_state)| index_state.block_number < block_number)
```

Comment:
It seems we can never index the genesis block. Users can have cells in the genesis block.